File: battleship.py
"""Battleship game module for Twoplay.

Variable grid size (8x8 .. 15x15) chosen at room creation, with a fleet that
scales by grid size. Manual placement, two turn rules ("strict" alternation or
"hit_again"). Firing reveals only hit/miss/sunk for the targeted cell.

Operates on a generic room's ``state`` / ``config`` dicts; the shared layer
(main.py) owns rooms, connections, reconnect and play-again bookkeeping.
"""

import random

from shared import ROLES, broadcast, opponent_role, send

GAME_TYPE = "battleship"

MIN_GRID = 8
MAX_GRID = 15
TURN_RULES = ("strict", "hit_again")

# Fleet table keyed by grid size N. Each entry: (ship name, length). Ship
# density stays roughly constant (~13-19%) as the board grows. Easy to tweak.
FLEETS: dict[int, list[tuple[str, int]]] = {
    8:  [("Battleship", 4), ("Cruiser", 3), ("Submarine", 3), ("Destroyer", 2)],
    9:  [("Battleship", 4), ("Cruiser", 3), ("Submarine", 3), ("Destroyer", 2), ("Destroyer", 2)],
    10: [("Carrier", 5), ("Battleship", 4), ("Cruiser", 3), ("Submarine", 3), ("Destroyer", 2)],
    11: [("Carrier", 5), ("Battleship", 4), ("Cruiser", 3), ("Submarine", 3), ("Destroyer", 2), ("Destroyer", 2)],
    12: [("Carrier", 5), ("Battleship", 4), ("Battleship", 4), ("Cruiser", 3), ("Submarine", 3), ("Destroyer", 2)],
    13: [("Carrier", 5), ("Battleship", 4), ("Battleship", 4), ("Cruiser", 3), ("Cruiser", 3), ("Submarine", 3), ("Destroyer", 2)],
    14: [("Carrier", 5), ("Carrier", 5), ("Battleship", 4), ("Battleship", 4), ("Cruiser", 3), ("Cruiser", 3), ("Destroyer", 2)],
    15: [("Carrier", 5), ("Carrier", 5), ("Battleship", 4), ("Battleship", 4), ("Cruiser", 3), ("Cruiser", 3), ("Submarine", 3), ("Destroyer", 2)],
}


def fleet_spec(grid_size: int) -> list[dict]:
    """Return the fleet as a list of {id, name, label, length}.

    When a ship name appears more than once it is numbered ("Destroyer 1",
    "Destroyer 2") so placement and sunk announcements stay unambiguous.
    """
    raw = FLEETS[grid_size]
    name_counts: dict[str, int] = {}
    for name, _ in raw:
        name_counts[name] = name_counts.get(name, 0) + 1
    seen: dict[str, int] = {}
    spec = []
    for idx, (name, length) in enumerate(raw):
        if name_counts[name] > 1:
            seen[name] = seen.get(name, 0) + 1
            label = f"{name} {seen[name]}"
        else:
            label = name
        spec.append({"id": f"ship{idx}", "name": name, "label": label, "length": length})
    return spec


# ---------------------------------------------------------------------------
# Room lifecycle hooks
# ---------------------------------------------------------------------------
def validate_create(data: dict):
    grid_size = data.get("grid_size")
    turn_rule = data.get("turn_rule")
    if not isinstance(grid_size, int) or not (MIN_GRID <= grid_size <= MAX_GRID):
        return None, f"Grid size must be {MIN_GRID}-{MAX_GRID}"
    if turn_rule not in TURN_RULES:
        return None, "Invalid turn rule"
    return {"grid_size": grid_size, "turn_rule": turn_rule}, None


def init_state(config: dict) -> dict:
    return {
        "fleet": fleet_spec(config["grid_size"]),
        "ships": {},            # role -> list of {id,label,length,cells:[[r,c]],hits:[[r,c]]}
        "ready": {},            # role -> bool (placement locked in)
        "shots": {"A": [], "B": []},  # role -> shots THAT role has fired
        "turn": None,
    }


def join_payload(room: dict) -> dict:
    return {
        "grid_size": room["config"]["grid_size"],
        "turn_rule": room["config"]["turn_rule"],
        "fleet": room["state"]["fleet"],
    }


def build_state_sync(room: dict, role: str) -> dict:
    state = room["state"]
    opp = opponent_role(role)
    opp_player = room["players"].get(opp)
    snapshot = {
        "type": "state_sync",
        "game_type": GAME_TYPE,
        "room_code": room["code"],
        "role": role,
        "grid_size": room["config"]["grid_size"],
        "turn_rule": room["config"]["turn_rule"],
        "fleet": state["fleet"],
        "your_ships": state["ships"].get(role, []),
        "you_ready": bool(state["ready"].get(role)),
        "opponent_present": opp_player is not None,
        "opponent_ready": bool(state["ready"].get(opp)),
        "opponent_connected": bool(opp_player and opp_player["connected"]),
        "started": room["started"],
        "over": room["over"],
        "turn": state["turn"],
        "your_shots": state["shots"].get(role, []),      # shots you fired
        "incoming_shots": state["shots"].get(opp, []),   # shots fired at you
        "winner": room["winner"],
    }
    if room["over"]:
        # Reveal both fleets so the boards can be shown in full.
        snapshot["boards"] = {r: state["ships"].get(r, []) for r in ROLES}
    return snapshot


def reset(room: dict, data: dict) -> dict:
    """Reset for another round. Players re-place their fleets."""
    config = room["config"]
    room["state"] = init_state(config)
    return {
        "type": "room_reset",
        "grid_size": config["grid_size"],
        "turn_rule": config["turn_rule"],
        "fleet": room["state"]["fleet"],
    }


def before_replay(room: dict, data: dict) -> None:
    # No per-round config changes for Battleship.
    return None


# ---------------------------------------------------------------------------
# Placement validation
# ---------------------------------------------------------------------------
def _cells_are_line(cells: list, length: int, grid_size: int) -> bool:
    """A valid ship: `length` distinct, in-bounds cells forming one straight,
    contiguous horizontal or vertical line."""
    if len(cells) != length:
        return False
    for r, c in cells:
        if not (isinstance(r, int) and isinstance(c, int)):
            return False
        if not (0 <= r < grid_size and 0 <= c < grid_size):
            return False
    rows = {r for r, c in cells}
    cols = {c for r, c in cells}
    if len(rows) == 1:  # horizontal
        cs = sorted(c for r, c in cells)
        return cs == list(range(cs[0], cs[0] + length))
    if len(cols) == 1:  # vertical
        rs = sorted(r for r, c in cells)
        return rs == list(range(rs[0], rs[0] + length))
    return False


def validate_placement(fleet: list, placed: list, grid_size: int):
    """Validate a full fleet placement. Returns (ships, error).

    `placed` is the client payload: a list of {id, cells:[[r,c],...]}. Each
    fleet ship must be placed exactly once, as a contiguous line of the right
    length, with no overlap between ships.
    """
    if not isinstance(placed, list) or len(placed) != len(fleet):
        return None, "Place your whole fleet."
    by_id = {}
    for item in placed:
        if not isinstance(item, dict):
            return None, "Bad placement data."
        by_id[item.get("id")] = item.get("cells")

    occupied = set()
    ships = []
    for spec in fleet:
        cells = by_id.get(spec["id"])
        if cells is None or not isinstance(cells, list):
            return None, f"Missing placement for {spec['label']}."
        cells = [[int(r), int(c)] for r, c in cells] if _all_pairs(cells) else None
        if cells is None or not _cells_are_line(cells, spec["length"], grid_size):
            return None, f"{spec['label']} is placed incorrectly."
        for cell in cells:
            key = (cell[0], cell[1])
            if key in occupied:
                return None, "Ships may not overlap."
            occupied.add(key)
        ships.append(
            {
                "id": spec["id"],
                "label": spec["label"],
                "length": spec["length"],
                "cells": cells,
                "hits": [],
            }
        )
    return ships, None


def _all_pairs(cells) -> bool:
    return all(
        isinstance(p, (list, tuple)) and len(p) == 2 and all(isinstance(x, int) for x in p)
        for p in cells
    )


# ---------------------------------------------------------------------------
# Firing
# ---------------------------------------------------------------------------
def _next_turn(shooter: str, result: str, turn_rule: str) -> str:
    if result in ("hit", "sunk") and turn_rule == "hit_again":
        return shooter
    return opponent_role(shooter)


# ---------------------------------------------------------------------------
# Message handler
# ---------------------------------------------------------------------------
async def handle(ws, room: dict, role: str, data: dict) -> None:
    msg_type = data.get("type")
    state = room["state"]
    config = room["config"]
    grid_size = config["grid_size"]

    if msg_type == "place_ships":
        if room["started"]:
            await send(ws, {"type": "error", "message": "Game already started"})
            return
        ships, err = validate_placement(state["fleet"], data.get("ships"), grid_size)
        if err:
            await send(ws, {"type": "error", "message": err})
            return
        state["ships"][role] = ships
        state["ready"][role] = True
        await send(ws, {"type": "placement_accepted"})
        opp = opponent_role(role)
        opp_player = room["players"].get(opp)
        if opp_player and opp_player["ws"]:
            await send(opp_player["ws"], {"type": "opponent_ready"})

        both_ready = (
            len(room["players"]) == 2
            and all(state["ready"].get(r) for r in room["players"])
        )
        if both_ready:
            room["started"] = True
            room["over"] = False
            state["turn"] = random.choice(ROLES)
            await broadcast(room, {"type": "game_start", "first_turn": state["turn"]})

    elif msg_type == "fire":
        if not room["started"] or room["over"] or state["turn"] != role:
            await send(ws, {"type": "error", "message": "Not your turn"})
            return
        row = data.get("row")
        col = data.get("col")
        if not (isinstance(row, int) and isinstance(col, int)):
            await send(ws, {"type": "error", "message": "Invalid target"})
            return
        if not (0 <= row < grid_size and 0 <= col < grid_size):
            await send(ws, {"type": "error", "message": "Target off the board"})
            return
        if any(s["row"] == row and s["col"] == col for s in state["shots"][role]):
            await send(ws, {"type": "error", "message": "Already fired there"})
            return

        defender = opponent_role(role)
        result = "miss"
        sunk_ship = None
        for ship in state["ships"][defender]:
            if [row, col] in ship["cells"]:
                result = "hit"
                if [row, col] not in ship["hits"]:
                    ship["hits"].append([row, col])
                if len(ship["hits"]) == ship["length"]:
                    result = "sunk"
                    sunk_ship = ship["label"]
                break

        state["shots"][role].append(
            {"row": row, "col": col, "result": result, "sunk_ship": sunk_ship}
        )

        all_sunk = all(
            len(s["hits"]) == s["length"] for s in state["ships"][defender]
        )
        if all_sunk:
            room["over"] = True
            state["turn"] = None
            room["winner"] = role
            await broadcast(
                room,
                {
                    "type": "fire_result",
                    "shooter": role,
                    "row": row,
                    "col": col,
                    "result": result,
                    "sunk_ship": sunk_ship,
                    "next_turn": None,
                },
            )
            await broadcast(
                room,
                {
                    "type": "game_over",
                    "winner": role,
                    "boards": {r: state["ships"].get(r, []) for r in ROLES},
                },
            )
        else:
            nxt = _next_turn(role, result, config["turn_rule"])
            state["turn"] = nxt
            await broadcast(
                room,
                {
                    "type": "fire_result",
                    "shooter": role,
                    "row": row,
                    "col": col,
                    "result": result,
                    "sunk_ship": sunk_ship,
                    "next_turn": nxt,
                },
            )

    else:
        await send(ws, {"type": "error", "message": "Unknown message type"})

File: static/battleship.js
// Battleship — game controller, built on the shared Twoplay networking layer.
(function () {
  "use strict";

  const $ = Twoplay.$;
  const GAME_TYPE = "battleship";

  const state = {
    roomCode: null,
    role: null,
    token: null,
    gridSize: 10,
    turnRule: "strict",
    fleet: [],          // [{id, name, label, length}]
    // Placement
    placed: {},         // id -> [[r,c], ...]
    currentShipId: null,
    orientation: "H",   // "H" | "V"
    ready: false,
    // Live game
    started: false,
    gameOver: false,
    turn: null,
    myTurn: false,
    shipCells: new Set(),   // my own ship cells, as "r,c"
    myShots: [],            // shots I fired:   {row,col,result,sunk_ship}
    incomingShots: [],      // shots fired at me:{row,col,result,sunk_ship}
  };

  // ---- Shared networking ------------------------------------------------
  const net = Twoplay.createNet(GAME_TYPE, {
    onMessage: handleMessage,
    shouldReconnect: () => !state.gameOver,
    onReconnecting: () => Twoplay.banner("Connection lost — reconnecting…", "warn"),
    onConnectionLost: () => {
      if (!state.gameOver) Twoplay.banner("Connection lost. Refresh to start over.", "error");
    },
  });

  function saveSession() {
    if (state.roomCode && state.role && state.token) {
      net.saveSession({
        roomCode: state.roomCode,
        role: state.role,
        token: state.token,
      });
    }
  }

  // ---- Screens ----------------------------------------------------------
  const screens = {
    home: $("screen-home"),
    lobby: $("screen-lobby"),
    place: $("screen-place"),
    game: $("screen-game"),
  };

  function showScreen(name) {
    Object.values(screens).forEach((s) => s.classList.add("hidden"));
    screens[name].classList.remove("hidden");
    $("btn-newgame").classList.toggle("hidden", name === "home");
  }

  // ---- Board building ---------------------------------------------------
  function cellSize(size) {
    return Math.max(16, Math.min(34, Math.floor(340 / size)));
  }

  function buildBoard(container, size, onCellClick) {
    container.innerHTML = "";
    const cs = cellSize(size);
    container.style.setProperty("--cell", cs + "px");
    container.style.gridTemplateColumns = `${cs}px repeat(${size}, ${cs}px)`;

    // Header row: empty corner + column letters.
    const corner = document.createElement("div");
    corner.className = "board-corner";
    container.appendChild(corner);
    for (let c = 0; c < size; c++) {
      const head = document.createElement("div");
      head.className = "board-head";
      head.textContent = String.fromCharCode(65 + c);
      container.appendChild(head);
    }

    const cells = [];
    for (let r = 0; r < size; r++) {
      cells.push([]);
      const num = document.createElement("div");
      num.className = "board-rownum";
      num.textContent = r + 1;
      container.appendChild(num);
      for (let c = 0; c < size; c++) {
        const cell = document.createElement("div");
        cell.className = "cell";
        cell.dataset.r = r;
        cell.dataset.c = c;
        if (onCellClick) {
          cell.addEventListener("click", () => onCellClick(r, c));
        }
        cells[r].push(cell);
        container.appendChild(cell);
      }
    }
    container._cells = cells;
    return cells;
  }

  const getCell = (board, r, c) => board._cells && board._cells[r] && board._cells[r][c];

  // ---- Placement --------------------------------------------------------
  function initPlacement() {
    state.placed = {};
    state.ready = false;
    state.orientation = "H";
    state.currentShipId = state.fleet.length ? state.fleet[0].id : null;
    updateRotateLabel();
    $("place-error").classList.add("hidden");
    $("place-status").classList.add("hidden");
    $("btn-ready").disabled = true;
    buildBoard($("place-board"), state.gridSize, onPlaceCell);
    renderFleetList();
    renderPlaceBoard();
    showScreen("place");
  }

  function shipById(id) {
    return state.fleet.find((s) => s.id === id);
  }

  function computeShipCells(ship, r, c) {
    const cells = [];
    for (let i = 0; i < ship.length; i++) {
      cells.push(state.orientation === "H" ? [r, c + i] : [r + i, c]);
    }
    return cells;
  }

  function inBounds(cells) {
    return cells.every(([r, c]) => r >= 0 && r < state.gridSize && c >= 0 && c < state.gridSize);
  }

  function onPlaceCell(r, c) {
    const ship = shipById(state.currentShipId);
    if (!ship) return;
    const cells = computeShipCells(ship, r, c);
    if (!inBounds(cells)) {
      showPlaceError(`${ship.label} won't fit there — try another cell or rotate.`);
      return;
    }
    // Overlap check against every OTHER placed ship.
    const occupied = new Set();
    for (const [id, cs] of Object.entries(state.placed)) {
      if (id === ship.id) continue;
      cs.forEach(([rr, cc]) => occupied.add(rr + "," + cc));
    }
    if (cells.some(([rr, cc]) => occupied.has(rr + "," + cc))) {
      showPlaceError("Ships may not overlap.");
      return;
    }
    state.placed[ship.id] = cells;
    $("place-error").classList.add("hidden");
    // Advance to the next unplaced ship, if any.
    const next = state.fleet.find((s) => !state.placed[s.id]);
    if (next) state.currentShipId = next.id;
    renderFleetList();
    renderPlaceBoard();
    $("btn-ready").disabled = Object.keys(state.placed).length !== state.fleet.length;
  }

  function showPlaceError(msg) {
    const e = $("place-error");
    e.textContent = msg;
    e.classList.remove("hidden");
  }

  function renderFleetList() {
    const list = $("fleet-list");
    list.innerHTML = "";
    state.fleet.forEach((ship) => {
      const chip = document.createElement("button");
      chip.type = "button";
      chip.className = "fleet-chip";
      if (state.placed[ship.id]) chip.classList.add("placed");
      if (ship.id === state.currentShipId) chip.classList.add("active");
      const name = document.createElement("span");
      name.className = "fleet-name";
      name.textContent = ship.label;
      const cells = document.createElement("span");
      cells.className = "fleet-cells";
      for (let i = 0; i < ship.length; i++) {
        const s = document.createElement("span");
        s.className = "fleet-cell";
        cells.appendChild(s);
      }
      chip.appendChild(name);
      chip.appendChild(cells);
      chip.addEventListener("click", () => {
        state.currentShipId = ship.id;
        renderFleetList();
        renderPlaceBoard();
      });
      list.appendChild(chip);
    });
  }

  function renderPlaceBoard() {
    const board = $("place-board");
    if (!board._cells) return;
    board._cells.flat().forEach((cell) => (cell.className = "cell"));
    for (const [id, cells] of Object.entries(state.placed)) {
      const current = id === state.currentShipId;
      cells.forEach(([r, c]) => {
        const cell = getCell(board, r, c);
        if (cell) {
          cell.classList.add("ship");
          if (current) cell.classList.add("ship-current");
        }
      });
    }
  }

  function rotate() {
    state.orientation = state.orientation === "H" ? "V" : "H";
    updateRotateLabel();
  }
  function updateRotateLabel() {
    $("btn-rotate").textContent =
      "↻ Rotate: " + (state.orientation === "H" ? "Horizontal" : "Vertical");
  }

  function submitPlacement() {
    if (Object.keys(state.placed).length !== state.fleet.length) return;
    const ships = state.fleet.map((s) => ({ id: s.id, cells: state.placed[s.id] }));
    state.shipCells = new Set();
    ships.forEach((s) => s.cells.forEach(([r, c]) => state.shipCells.add(r + "," + c)));
    net.send({ type: "place_ships", ships });
    state.ready = true;
    $("btn-ready").disabled = true;
    $("place-status").classList.remove("hidden");
  }

  // ---- Live game --------------------------------------------------------
  function startGame(firstTurn) {
    state.started = true;
    state.gameOver = false;
    state.myShots = [];
    state.incomingShots = [];
    $("sunk-feed").innerHTML = "";
    $("game-result").classList.add("hidden");
    $("play-again-status").classList.add("hidden");
    buildBoard($("fire-board"), state.gridSize, onFireCell);
    buildBoard($("own-board"), state.gridSize, null);
    renderFireBoard();
    renderOwnBoard();
    setTurn(firstTurn);
    showScreen("game");
  }

  function setTurn(turn) {
    state.turn = turn;
    state.myTurn = turn === state.role;
    const banner = $("turn-banner");
    const fire = $("fire-board");
    if (state.myTurn) {
      banner.textContent = "Your turn — fire a shot!";
      banner.className = "turn-banner my-turn";
      fire.classList.remove("inactive");
    } else {
      banner.textContent = "Opponent's turn — waiting…";
      banner.className = "turn-banner their-turn";
      fire.classList.add("inactive");
    }
  }

  function onFireCell(r, c) {
    if (!state.myTurn || state.gameOver) return;
    if (state.myShots.some((s) => s.row === r && s.col === c)) return;
    net.send({ type: "fire", row: r, col: c });
  }

  function renderFireBoard() {
    const board = $("fire-board");
    if (!board._cells) return;
    board._cells.flat().forEach((cell) => (cell.className = "cell"));
    state.myShots.forEach((s) => {
      const cell = getCell(board, s.row, s.col);
      if (cell) cell.classList.add(s.result === "miss" ? "miss" : "hit", "fired");
      if (cell && s.result === "sunk") cell.classList.add("sunk");
    });
  }

  function renderOwnBoard() {
    const board = $("own-board");
    if (!board._cells) return;
    board._cells.flat().forEach((cell) => (cell.className = "cell"));
    state.shipCells.forEach((key) => {
      const [r, c] = key.split(",").map(Number);
      const cell = getCell(board, r, c);
      if (cell) cell.classList.add("ship");
    });
    state.incomingShots.forEach((s) => {
      const cell = getCell(board, s.row, s.col);
      if (cell) cell.classList.add(s.result === "miss" ? "miss" : "hit");
    });
  }

  function revealOpponentFleet(oppShips) {
    const board = $("fire-board");
    if (!board._cells || !oppShips) return;
    oppShips.forEach((ship) => {
      ship.cells.forEach(([r, c]) => {
        const cell = getCell(board, r, c);
        if (cell && !cell.classList.contains("hit")) cell.classList.add("ship-revealed");
      });
    });
  }

  function announceSunk(shooter, label) {
    if (!label) return;
    const div = document.createElement("div");
    div.className = "sunk-msg";
    if (shooter === state.role) {
      div.classList.add("good");
      div.textContent = `💥 You sank the opponent's ${label}!`;
    } else {
      div.classList.add("bad");
      div.textContent = `🔥 Your ${label} was sunk!`;
    }
    const feed = $("sunk-feed");
    feed.insertBefore(div, feed.firstChild);
    while (feed.children.length > 4) feed.removeChild(feed.lastChild);
  }

  // ---- Message handling -------------------------------------------------
  function handleMessage(msg) {
    switch (msg.type) {
      case "room_created":
        state.roomCode = msg.room_code;
        state.role = msg.player_role || "A";
        state.token = msg.token;
        applyConfig(msg);
        saveSession();
        $("lobby-roomcode").textContent = msg.room_code;
        $("lobby-status").textContent = "Waiting for an opponent to join…";
        showScreen("lobby");
        break;

      case "waiting_for_opponent":
        break;

      case "joined_room":
        state.roomCode = msg.room_code;
        state.role = msg.player_role;
        state.token = msg.token;
        applyConfig(msg);
        saveSession();
        initPlacement();
        break;

      case "opponent_joined":
        // Creator's opponent arrived — both move to fleet placement.
        initPlacement();
        break;

      case "placement_accepted":
        state.ready = true;
        $("place-status").classList.remove("hidden");
        break;

      case "opponent_ready":
        if (!screens.place.classList.contains("hidden")) {
          $("place-hint").textContent = "Opponent is ready and waiting for you.";
        }
        break;

      case "game_start":
        startGame(msg.first_turn);
        break;

      case "fire_result": {
        const shot = {
          row: msg.row,
          col: msg.col,
          result: msg.result,
          sunk_ship: msg.sunk_ship,
        };
        if (msg.shooter === state.role) {
          state.myShots.push(shot);
          renderFireBoard();
        } else {
          state.incomingShots.push(shot);
          renderOwnBoard();
        }
        announceSunk(msg.shooter, msg.sunk_ship);
        if (msg.next_turn) setTurn(msg.next_turn);
        break;
      }

      case "game_over":
        endGame(msg.winner, msg.boards);
        break;

      case "state_sync":
        rebuildFromState(msg);
        break;

      case "rejoin_failed":
        net.clearSession();
        Twoplay.hideBanner();
        Twoplay.banner("Your previous game is no longer available.", "warn");
        showScreen("home");
        break;

      case "opponent_reconnected":
        Twoplay.hideBanner();
        break;

      case "opponent_disconnected":
        Twoplay.banner("Your opponent disconnected — they can rejoin anytime.", "warn");
        break;

      case "room_reset":
        applyConfig(msg);
        // Clear the finished-game flags so reconnect works during replay setup.
        state.gameOver = false;
        state.started = false;
        Twoplay.hideBanner();
        $("game-result").classList.add("hidden");
        $("play-again-status").classList.add("hidden");
        $("btn-play-again").disabled = false;
        $("place-hint").textContent =
          "Select a ship, choose its orientation, then click a cell to place it.";
        initPlacement();
        break;

      case "opponent_left":
        net.clearSession();
        Twoplay.banner("Your opponent left the game.", "warn");
        resetToGameHome();
        break;

      case "error":
        showError(msg.message);
        break;
    }
  }

  function applyConfig(msg) {
    if (msg.grid_size) state.gridSize = msg.grid_size;
    if (msg.turn_rule) state.turnRule = msg.turn_rule;
    if (msg.fleet) state.fleet = msg.fleet;
  }

  function endGame(winner, boards) {
    state.gameOver = true;
    state.turn = null;
    state.myTurn = false;
    const won = winner === state.role;
    const headline = $("over-headline");
    headline.textContent = won ? "🎉 Victory!" : "Defeated";
    headline.className = won ? "win" : "lose";
    $("over-detail").textContent = won
      ? "You sank the opponent's entire fleet — their ships are revealed on the left."
      : "Your fleet was wiped out. The opponent's ships are revealed on the left.";
    if (boards) {
      const oppRole = state.role === "A" ? "B" : "A";
      revealOpponentFleet(boards[oppRole]);
    }
    // Show the result in place so the final, revealed boards stay on screen.
    const banner = $("turn-banner");
    banner.textContent = won ? "🎉 Victory!" : "Defeated";
    banner.className = "turn-banner " + (won ? "my-turn" : "their-turn");
    $("fire-board").classList.add("inactive");
    $("btn-play-again").disabled = false;
    $("play-again-status").classList.add("hidden");
    $("game-result").classList.remove("hidden");
    showScreen("game");
  }

  function rebuildFromState(s) {
    state.roomCode = s.room_code;
    state.role = s.role;
    state.gridSize = s.grid_size;
    state.turnRule = s.turn_rule;
    state.fleet = s.fleet || [];
    state.started = s.started;
    state.gameOver = s.over;
    saveSession();
    Twoplay.hideBanner();

    // Rebuild my own ship cells from the server's record (if I'd placed).
    state.shipCells = new Set();
    (s.your_ships || []).forEach((ship) =>
      ship.cells.forEach(([r, c]) => state.shipCells.add(r + "," + c))
    );
    state.myShots = s.your_shots || [];
    state.incomingShots = s.incoming_shots || [];

    if (!s.started && !s.over && !s.opponent_present) {
      $("lobby-roomcode").textContent = state.roomCode;
      $("lobby-status").textContent = "Waiting for an opponent to join…";
      showScreen("lobby");
      return;
    }

    if (!s.started && !s.over) {
      // Placement phase.
      if (s.you_ready) {
        // Already locked in — show the waiting state with my placed fleet.
        buildBoard($("place-board"), state.gridSize, null);
        renderReadyPlaceBoard();
        $("btn-ready").disabled = true;
        $("place-status").classList.remove("hidden");
        showScreen("place");
      } else {
        initPlacement();
        if (s.opponent_ready) {
          $("place-hint").textContent = "Opponent is ready and waiting for you.";
        }
      }
      return;
    }

    if (s.over) {
      buildBoard($("fire-board"), state.gridSize, onFireCell);
      buildBoard($("own-board"), state.gridSize, null);
      renderFireBoard();
      renderOwnBoard();
      const oppRole = state.role === "A" ? "B" : "A";
      if (s.boards) revealOpponentFleet(s.boards[oppRole]);
      endGame(s.winner, s.boards);
      return;
    }

    // Game in progress.
    buildBoard($("fire-board"), state.gridSize, onFireCell);
    buildBoard($("own-board"), state.gridSize, null);
    renderFireBoard();
    renderOwnBoard();
    setTurn(s.turn);
    showScreen("game");
    if (!s.opponent_connected) {
      Twoplay.banner("Your opponent disconnected — they can rejoin anytime.", "warn");
    }
  }

  // Show my placed fleet on the placement board while waiting (read-only).
  function renderReadyPlaceBoard() {
    const board = $("place-board");
    if (!board._cells) return;
    state.shipCells.forEach((key) => {
      const [r, c] = key.split(",").map(Number);
      const cell = getCell(board, r, c);
      if (cell) cell.classList.add("ship");
    });
  }

  function showError(message) {
    if (!screens.place.classList.contains("hidden")) {
      showPlaceError(message);
    } else {
      Twoplay.banner(message, "error");
    }
  }

  // ---- Leaving ----------------------------------------------------------
  function resetToGameHome() {
    net.clearSession();
    net.leave();
    state.roomCode = null;
    state.role = null;
    state.token = null;
    state.started = false;
    state.gameOver = false;
    state.turn = null;
    state.ready = false;
    state.placed = {};
    state.shipCells = new Set();
    state.myShots = [];
    state.incomingShots = [];
    Twoplay.hideBanner();
    $("join-code").value = "";
    showScreen("home");
  }

  function newGame() {
    const inProgress = state.started && !state.gameOver;
    if (inProgress && !confirm("Leave this game and start a new one?")) return;
    net.send({ type: "leave_room" });
    resetToGameHome();
  }

  function goToArcadeHome(e) {
    if (e) e.preventDefault();
    const inProgress = state.started && !state.gameOver;
    if (inProgress && !confirm("Leave this game and return to Twoplay home?")) return;
    net.send({ type: "leave_room" });
    net.clearSession();
    setTimeout(() => (location.href = "/"), 60);
  }

  // ---- Wire up ----------------------------------------------------------
  $("btn-create").addEventListener("click", () => {
    const grid_size = parseInt($("grid-size").value, 10);
    const turn_rule = $("turn-rule").value;
    net.connect();
    net.send({ type: "create_room", game_type: GAME_TYPE, grid_size, turn_rule });
  });

  $("btn-join").addEventListener("click", () => {
    const code = $("join-code").value.trim().toUpperCase();
    if (!code) {
      Twoplay.banner("Enter a room code to join.", "error");
      return;
    }
    net.connect();
    net.send({ type: "join_room", game_type: GAME_TYPE, room_code: code });
  });

  $("join-code").addEventListener("keydown", (e) => {
    if (e.key === "Enter") $("btn-join").click();
  });

  $("btn-copy").addEventListener("click", () => Twoplay.copyText(state.roomCode, $("btn-copy")));
  $("btn-rotate").addEventListener("click", rotate);
  $("btn-reset-ships").addEventListener("click", initPlacement);
  $("btn-ready").addEventListener("click", submitPlacement);
  $("btn-newgame").addEventListener("click", newGame);
  $("brand-home").addEventListener("click", goToArcadeHome);

  $("btn-play-again").addEventListener("click", () => {
    $("btn-play-again").disabled = true;
    $("play-again-status").classList.remove("hidden");
    net.send({ type: "play_again" });
  });

  // ---- Startup ----------------------------------------------------------
  const params = new URLSearchParams(location.search);
  const preRoom = params.get("room");
  if (preRoom) {
    net.clearSession();
    $("join-code").value = preRoom.toUpperCase();
    showScreen("home");
  } else if (net.loadSession()) {
    Twoplay.banner("Reconnecting to your game…", "warn");
    net.doReconnect();
  } else {
    showScreen("home");
  }
})();
